src/opm/parser/eclipse/EclipseState/Schedule/UDQ/UDQFunction.cpp

```diff
@@ -0,0 +1,509 @@
+/*
+  Copyright 2019 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <unordered_set>
+#include <cmath>
+#include <algorithm>
+#include <iostream>
+#include <random>
+
+#include <opm/parser/eclipse/EclipseState/Schedule/UDQ/UDQFunction.hpp>
+
+namespace Opm {
+
+UDQFunction::UDQFunction(const std::string& name) :
+        m_name(name)
+{
+}
+
+const std::string& UDQFunction::name() const {
+    return this->m_name;
+}
+
+UDQScalarFunction::UDQScalarFunction(const std::string&name, std::function<UDQScalar(const UDQSet& arg)> f) :
+    UDQFunction(name),
+    func(f)
+{
+}
+
+
+
+UDQScalar UDQScalarFunction::eval(const UDQSet& arg) const {
+    return this->func(arg);
+}
+
+
+UDQSet UDQUnaryElementalFunction::eval(const UDQSet& arg) const {
+    return this->func(arg);
+}
+
+/*****************************************************************/
+UDQScalar UDQScalarFunction::MIN(const UDQSet& arg) {
+    std::vector<double> defined_values = arg.defined_values();
+    return *std::min_element(defined_values.begin(), defined_values.end());
```

Comment:
This will dereference `.end()` if the sequence is empty.  Is it possible to call `MIN()` on an empty sequence of `defined_values()`?